File: fe/ClinicManager/src/models/voucher.ts
import { Client } from "./client";

export enum VoucherStatus {
    ACTIVE = "ACTIVE",
    USED = "USED",
    EXPIRED = "EXPIRED"
}

export interface Voucher {
    id: number;
    value: number;
    issueDate: string;
    expiryDate: string;
    client: Client;
    status: VoucherStatus;
    purchaseVisitId?: number;
    createdBy: number | null;
}

export interface NewVoucher {
    value?: number;
    client?: Client;
    issueDate?: string | null;
    expiryDate?: string | null;
}

export interface VoucherFilterDTO {
    status?: VoucherStatus | null;
    keyword?: string | null;
}

File: be/ClinicManager/src/main/java/com/clinic/clinicmanager/model/Voucher.java
package com.clinic.clinicmanager.model;

import com.clinic.clinicmanager.model.constants.VoucherStatus;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Table(name = "voucher")
public class Voucher {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Double value;

    @Column(nullable = false)
    private LocalDate issueDate;

    @Column(nullable = false)
    private LocalDate expiryDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "client_id", nullable = false)
    private Client client;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private VoucherStatus status = VoucherStatus.ACTIVE;

    @Column(name = "created_by_user_id")
    private Long createdByUserId;
}
